### PR TITLE
Pixel shuffle

### PR DESCRIPTION
**Pull request by AnnaAltynova**

### CODE REVIEW

File: batchflow/models/torch/layers/pixel_shuffle.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
+""" Layer implementing Pixel Shuffle for an arbitrary dimensionality.
```
Comment (SergeyTsimfer) [inline batchflow/models/torch/layers/pixel_shuffle.py:1]: add a word about PixelUnshuffle

File: batchflow/models/torch/layers/pixel_shuffle.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    ----------
+    upscale_factor : int
+        factor to increase spatial resolution by.
+    """
+    def __init__(self, upscale_factor):
+        super().__init__()
+        self.upscale_factor = upscale_factor
+
+    def forward(self, x):
+        batch_size, channels, *dims = x.size()                                         # (B, C * r^N, H, W)
```
Comment (SergeyTsimfer) [inline batchflow/models/torch/layers/pixel_shuffle.py:25]: `C * r * r` / `C * r^2` in this case

File: batchflow/models/torch/layers/pixel_shuffle.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        ndims = len(dims)
+        channels //= (self.upscale_factor**ndims)
```
Comment (SergeyTsimfer) [inline batchflow/models/torch/layers/pixel_shuffle.py:27]: do you need parenthesis here?

File: batchflow/models/torch/layers/pixel_shuffle.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+        factor_expand = [self.upscale_factor] * ndims
+        x = x.contiguous().view(                                                       # (B, C, r, r, H, W)
+            batch_size, channels, *factor_expand, *dims
+            )
```
Comment (SergeyTsimfer) [inline batchflow/models/torch/layers/pixel_shuffle.py:32]:
```suggestion
        x = x.contiguous().view(batch_size, channels,                                  # (B, C, r, r, H, W)
                                *factor_expand, *dims)
```

File: batchflow/models/torch/layers/pixel_shuffle.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+        permute = [None] * (2 * ndims)
+        permute[::2] = range(ndims + 2, 2 * ndims + 2)
+        permute[1::2] = range(2, ndims + 2)
+        x = x.permute(0, 1, *permute).contiguous()                                     # (B, C, H, r, W, r)
```
Comment (SergeyTsimfer) [inline batchflow/models/torch/layers/pixel_shuffle.py:37]: Do you need this call to `contiguous`?

File: batchflow/models/torch/layers/pixel_shuffle.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+Shi et al. "`Real-Time Single Image and Video Super-Resolution
+Using an Efficient Sub-Pixel Convolutional Neural Network_
+<https://arxiv.org/pdf/1609.05158.pdf>`"
+"""
+from torch import nn
+
+
+
+class PixelShuffle(nn.Module):
+    """ Rearranges elements in a tensor of shape [B, C * r^N, D1, D2, ... DN]
```
Comment (SergeyTsimfer) [inline batchflow/models/torch/layers/pixel_shuffle.py:11]: I do like the idea of copy-paste the documentation from `PyTorch` counterpart, but maybe something along the lines of `just as in PyTorch, but works with arbitrary number of dimensions` would be helpful

File: batchflow/models/torch/layers/pixel_shuffle.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+    def forward(self, x):
+        x = x.contiguous()                                                             # (B, C, H * r, B * r)
+        batch_size, channels, *dims = x.size()
+        ndims = len(dims)
+
+        out_dims = [dim // self.downscale_factor for dim in dims]
+        reshape = [None] * (2 * ndims)
+        reshape[::2] = out_dims
+        reshape[1::2] = [self.downscale_factor for i in range(ndims)]
```
Comment (SergeyTsimfer) [inline batchflow/models/torch/layers/pixel_shuffle.py:69]:
```python
reshape[1::2] = [self.downscale_factor for _ in ndims]
```
better yet:
```python
reshape[1::2] = [self.downscale_factor] * ndims
```

File: batchflow/models/torch/layers/pixel_shuffle.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+        x = x.view(batch_size, channels, *reshape)                                     # (B, C, H, r, W, r)
+
+        permute = [None] * (2 * ndims)
```
Comment (SergeyTsimfer) [inline batchflow/models/torch/layers/pixel_shuffle.py:73]: do you need those parenthesis?

File: batchflow/models/torch/layers/pixel_shuffle.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+        permute[:2 * ndims] = range(3, 2 * ndims + 2, 2)
+        permute[2 * ndims:] = range(2, 2 * ndims + 1, 2)
+        x = x.permute(0, 1, *permute).contiguous()                                     # (B, C, r, r, H, W)
```
Comment (SergeyTsimfer) [inline batchflow/models/torch/layers/pixel_shuffle.py:76]: do you need this call to `contiguous`

**Review by SergeyTsimfer — COMMENTED** [review]
Overall, looks nice

Please, make it so that these layers are used in the `batchflow.models.torch.Block` class in the layout convention

File: batchflow/models/torch/layers/pixel_shuffle.py [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+
+class PixelShuffle(nn.Module):
+    """ Rearranges elements in a tensor of shape [B, C * r^N, D1, D2, ... DN]
+    to a tensor of shape [B, C, D1 * r, D2 * r, ... DN * r]
+    where r is an upscale factor.
+
+    Parameters
+    ----------
+    upscale_factor : int
+        factor to increase spatial resolution by.
```
Comment (HollowPrincess) [inline batchflow/models/torch/layers/pixel_shuffle.py:18]:
```suggestion
        Factor to increase spatial resolution by.
```

File: batchflow/models/torch/layers/pixel_shuffle.py [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+    def forward(self, x):
+        batch_size, channels, *dims = x.size()                                         # (B, C * r^N, H, W)
+        ndims = len(dims)
+        channels //= (self.upscale_factor**ndims)
+
+        factor_expand = [self.upscale_factor] * ndims
+        x = x.contiguous().view(batch_size, channels,                                  # (B, C, r, r, H, W)
+                                *factor_expand, *dims)
+
+        permute = [None] * (2 * ndims)
```
Comment (HollowPrincess) [inline batchflow/models/torch/layers/pixel_shuffle.py:33]: Maybe `permute_dims`?

File: batchflow/models/torch/layers/pixel_shuffle.py [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+
+
+class PixelUnshuffle(nn.Module):
+    """ Rearranges elements in a tensor of shape [B, C, D1 * r, D2 * r, ... DN * r]
+    to a tensor of shape [B, C * r^N, D1, DN], where r is a downscale factor.
+
+    Parameters
+    ----------
+    downscale_factor : int
+        factor to decrease spatial resolution by.
```
Comment (HollowPrincess) [inline batchflow/models/torch/layers/pixel_shuffle.py:54]:
```suggestion
        Factor to decrease spatial resolution by.
```

File: batchflow/models/torch/layers/pixel_shuffle.py [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+        factor to increase spatial resolution by.
+    """
+    def __init__(self, upscale_factor):
+        super().__init__()
+        self.upscale_factor = upscale_factor
+
+    def forward(self, x):
+        batch_size, channels, *dims = x.size()                                         # (B, C * r^N, H, W)
+        ndims = len(dims)
+        channels //= (self.upscale_factor**ndims)
```
Comment (HollowPrincess) [inline batchflow/models/torch/layers/pixel_shuffle.py:27]: Maybe `out_channels`? It is not necessary, but looks clearer.

File: batchflow/models/torch/layers/pixel_shuffle.py [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+        super().__init__()
+        self.downscale_factor = downscale_factor
+
+    def forward(self, x):
+        x = x.contiguous()                                                             # (B, C, H * r, B * r)
+        batch_size, channels, *dims = x.size()
+        ndims = len(dims)
+
+        out_dims = [dim // self.downscale_factor for dim in dims]
+        reshape = [None] * (2 * ndims)
```
Comment (HollowPrincess) [inline batchflow/models/torch/layers/pixel_shuffle.py:66]: Maybe `shape_dims` or smth like that? It is not a `reshape`, and it is not a full `shape`.

File: batchflow/models/torch/layers/pixel_shuffle.py (hunk 7 above)
Comment (HollowPrincess) [reply batchflow/models/torch/layers/pixel_shuffle.py:68]: Why not in the proposed way?

File: batchflow/models/torch/layers/pixel_shuffle.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (AnnaAltynova) [reply batchflow/models/torch/layers/pixel_shuffle.py]: added ValueError

File: batchflow/models/torch/layers/pixel_shuffle.py [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+    ----------
+    downscale_factor : int
+        Factor to decrease spatial resolution by.
+    """
+    def __init__(self, downscale_factor):
+        super().__init__()
+        self.downscale_factor = downscale_factor
+
+    def forward(self, x):
+        x = x.contiguous()                                                             # (B, C, H * r, B * r)
```
Comment (SergeyTsimfer) [inline batchflow/models/torch/layers/pixel_shuffle.py:66]: type in the dimensions

**Review by SergeyTsimfer — APPROVED** [review]
